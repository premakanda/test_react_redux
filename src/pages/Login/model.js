// @flow

export type State = {|
  +user: ?string,
  +errorMsg: ?string,
  +isLoading: boolean,
|}
