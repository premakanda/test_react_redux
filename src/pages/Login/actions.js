// @flow

import { API_ROOT } from '../../constants/Defaults'
import { checkResponse } from '../../helpers/session'
import { postData } from '../../helpers/network'
import { defaultErrorMsg } from '../../constants/Defaults'
import * as t from './actionTypes'

type CbType = () => void
type Params = {|
  +email: string,
  +password: string,
|}

export const loginRequest = () => ({
  type: t.LOG_IN_REQUEST,
})

export const loginSuccess = (email: string) => ({
  type: t.LOG_IN_SUCCESS,
  payload: {
    email,
  },
})

export const loginFailure = (errorMsg: string = defaultErrorMsg): any => ({
  type: t.LOG_IN_FAILURE,
  payload: {
    errorMsg,
  },
  error: true,
})

// TODO: make acync AC flow typed
export function logIn(params: Params, cb: CbType): Function {
  return (dispatch: any) => {
    dispatch(loginRequest())
    return postData(`${API_ROOT}/validate`, params)
      .then(res => {
        if (checkResponse(res)) {
          //TODO: smth from response
          dispatch(loginSuccess(params.email))
          cb()
        } else {
          dispatch(loginFailure(res.message))
        }
      })
      .catch(error => {
        dispatch(loginFailure())
      })
  }
}

export function logOut() {
  return {
    type: t.LOG_OUT,
  }
}
