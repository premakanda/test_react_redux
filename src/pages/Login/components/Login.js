import React from 'react'
import PropTypes from 'prop-types'
import { Redirect } from 'react-router-dom'
import { testPassword, testEmail } from '../../../helpers/inputs'

class Login extends React.Component {
  state = {
    redirectToPreviousRoute: false,
    email: '',
    password: '',
  }

  handleSubmit = e => {
    e.preventDefault()
    const { email, password } = this.state

    this.props.logIn(
      {
        email,
        password,
      },
      () => {
        this.setState({ redirectToPreviousRoute: true })
      }
    )
  }

  handleChange = e => {
    const value = e.currentTarget.value
    const fieldName = e.currentTarget.dataset.fieldName

    this.setState(prev => ({
      ...prev,
      [fieldName]: value,
    }))
  }

  validate = () => {
    const { email, password } = this.state
    if (!testPassword(password)) {
      return false
    }
    if (!testEmail(email)) {
      return false
    }
    return true
  }

  render() {
    const { location, errorMsg } = this.props
    const { from } = location.state || { from: { pathname: '/' } }
    const { email, password, redirectToPreviousRoute } = this.state

    if (redirectToPreviousRoute) {
      return <Redirect to={from} />
    }

    return (
      <div>
        {errorMsg && <p>{errorMsg}</p>}
        <form onSubmit={this.handleSubmit}>
          <input
            data-field-name={'email'}
            type={'text'}
            onChange={this.handleChange}
            placeholder={'e-mail'}
            value={email}
          />
          <input
            data-field-name={'password'}
            type={'text'}
            onChange={this.handleChange}
            placeholder={'Пароль'}
            value={password}
          />
          <button disabled={!this.validate()} type="submit">
            Log in
          </button>
        </form>
      </div>
    )
  }
}

Login.propTypes = {
  logIn: PropTypes.func.isRequired,
  errorMsg: PropTypes.string,
}

export default Login
