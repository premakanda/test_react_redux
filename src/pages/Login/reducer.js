//@ flow

import * as t from './actionTypes'
import type { State } from './model'

export const initialState: State = {
  user: null,
  errorMsg: null,
  isLoading: false,
}

export default (state: State = initialState, action: any): State => {
  switch (action.type) {
    case t.LOG_IN_REQUEST:
      return {
        ...state,
        isLoading: true,
        errorMsg: null,
      }
    case t.LOG_IN_SUCCESS:
      return {
        ...state,
        user: {
          name: action.payload.email,
        },
        errorMsg: null,
      }
    case t.LOG_OUT:
      return {
        ...state,
        user: null,
        errorMsg: null,
      }
    case t.LOG_IN_FAILURE:
      return {
        ...state,
        errorMsg: action.payload.errorMsg,
      }
    default:
      return state
  }
}
