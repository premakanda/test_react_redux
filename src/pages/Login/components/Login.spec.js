import React from 'react'
import { shallow } from 'enzyme'
import Login from './Login'

describe('Login', () => {
  const mockLogin = jest.fn()

  const initialState = {
    email: '',
    password: '',
    redirectToPreviousRoute: false,
  }

  const props = {
    location: { state: { from: '/' } },
    logIn: mockLogin,
    errorMsg: null,
  }
  const login = shallow(<Login {...props} />)

  it('renders properly', () => {
    expect(login).toMatchSnapshot()
  })

  it('initialize Login with initial state', () => {
    expect(login.state()).toEqual(initialState)
  })

  describe('Form handlers', () => {
    describe('when typing into email input', () => {
      const email = 'max@test.com'

      beforeEach(() => {
        login.find('[data-field-name="email"]').simulate('change', {
          currentTarget: {
            value: email,
            dataset: {
              fieldName: 'email',
            },
          },
        })
      })

      it('updates email field in state', () => {
        expect(login.state().email).toEqual(email)
      })
    })
    describe('when typing into password input', () => {
      const password = '11'

      beforeEach(() => {
        login.find('[data-field-name="password"]').simulate('change', {
          currentTarget: {
            value: password,
            dataset: {
              fieldName: 'password',
            },
          },
        })
      })

      it('updates password field in state', () => {
        expect(login.state().password).toEqual(password)
      })
    })

    describe('when submiting the form', () => {
      beforeEach(() => {
        login.find('form').simulate('submit', {
          preventDefault: () => {},
        })
      })

      it('calls the props.logIn', () => {
        /* expect(mockLogin).toHaveBeenCalledWith(
          {
            email: 'max@test.com',
            password: '11',
          }
        ) */
        expect(mockLogin).toHaveBeenCalledTimes(1)
      })
    })
    describe('when clicking the Login button', () => {
      login.find('button').simulate('click', {
        preventDefault: () => {},
      })

      it('calls the props.logIn', () => {
        expect(mockLogin).toHaveBeenCalledTimes(1)
      })
    })

    afterAll(() => {
      login.setState(initialState)
    })
  })

  describe('disabled attr for submit button', () => {
    it('render form with disabled button at initial render', () => {
      const button = login.find('button')
      expect(button.prop('disabled')).toBeTruthy()
    })

    it('render form with disabled button for bad password', () => {
      const nextState = {
        ...initialState,
        password: 123,
        email: 'max@test.com',
      }
      login.setState(nextState)

      const button = login.find('button')
      expect(button.prop('disabled')).toBeTruthy()
    })

    it('render form with disabled button for bad email', () => {
      const nextState = {
        ...initialState,
        password: 12345,
        email: 'bademail',
      }
      login.setState(nextState)

      const button = login.find('button')
      expect(button.prop('disabled')).toBeTruthy()
    })

    it('render form with disabled button for bad email & password', () => {
      const nextState = {
        ...initialState,
        password: 1,
        email: 'bademail',
      }
      login.setState(nextState)

      const button = login.find('button')
      expect(button.prop('disabled')).toBeTruthy()
    })

    it('render form without disabled button, because email & password are correct', () => {
      const nextState = {
        ...initialState,
        email: 'max@test.com',
        password: '123456',
      }
      login.setState(nextState)

      const button = login.find('button')
      expect(button.prop('disabled')).toBeFalsy()
    })

    afterEach(() => {
      login.setState(initialState)
    })
  })
})
