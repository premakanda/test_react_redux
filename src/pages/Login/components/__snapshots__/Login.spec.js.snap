// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    errorMsg={null}
    location={
      Object {
        "state": Object {
          "from": "/",
        },
      }
    }
    logIn={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        null,
        <form
          onSubmit={[Function]}
        >
          <input
            data-field-name="email"
            onChange={[Function]}
            placeholder="e-mail"
            type="text"
            value=""
          />
          <input
            data-field-name="password"
            onChange={[Function]}
            placeholder="Пароль"
            type="text"
            value=""
          />
          <button
            disabled={true}
            type="submit"
          >
            Log in
          </button>
        </form>,
      ],
    },
    "ref": null,
    "rendered": Array [
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              data-field-name="email"
              onChange={[Function]}
              placeholder="e-mail"
              type="text"
              value=""
            />,
            <input
              data-field-name="password"
              onChange={[Function]}
              placeholder="Пароль"
              type="text"
              value=""
            />,
            <button
              disabled={true}
              type="submit"
            >
              Log in
            </button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "data-field-name": "email",
              "onChange": [Function],
              "placeholder": "e-mail",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "data-field-name": "password",
              "onChange": [Function],
              "placeholder": "Пароль",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Log in",
              "disabled": true,
              "type": "submit",
            },
            "ref": null,
            "rendered": "Log in",
            "type": "button",
          },
        ],
        "type": "form",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          null,
          <form
            onSubmit={[Function]}
          >
            <input
              data-field-name="email"
              onChange={[Function]}
              placeholder="e-mail"
              type="text"
              value=""
            />
            <input
              data-field-name="password"
              onChange={[Function]}
              placeholder="Пароль"
              type="text"
              value=""
            />
            <button
              disabled={true}
              type="submit"
            >
              Log in
            </button>
          </form>,
        ],
      },
      "ref": null,
      "rendered": Array [
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                data-field-name="email"
                onChange={[Function]}
                placeholder="e-mail"
                type="text"
                value=""
              />,
              <input
                data-field-name="password"
                onChange={[Function]}
                placeholder="Пароль"
                type="text"
                value=""
              />,
              <button
                disabled={true}
                type="submit"
              >
                Log in
              </button>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "data-field-name": "email",
                "onChange": [Function],
                "placeholder": "e-mail",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "data-field-name": "password",
                "onChange": [Function],
                "placeholder": "Пароль",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Log in",
                "disabled": true,
                "type": "submit",
              },
              "ref": null,
              "rendered": "Log in",
              "type": "button",
            },
          ],
          "type": "form",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
